### ISSUE TITLE
[torch.vmap] Support second order grads

### ISSUE THREAD

**Issue by n-gao** [pytorch/pytorch#53903]
## 🐛 Bug
`torch.vmap` could be a great tool to compute only certain parts of higher-order gradients, e.g., the Hessian.
A common use case would be to compute the Laplacian where only the diagonal of the Hessian is needed.
While this can be accomplished vectorized with `torch.autograd.functional.jacobian` and taking the derivative
The memory scales in `O(N²)`. A `torch.vmap` implementation would only require `O(N)` time and space.

## To Reproduce

```python
N = 5
f = lambda x: (x**2).sum()
x = torch.randn(N, requires_grad=True)
y = f(x)
df_dx = torch.autograd.grad(y.sum(), x, create_graph=True)

def diag_grad(grad_i, x_i):
    return torch.autograd.grad(grad_i, x_i)
d2f_dx2 = torch.vmap(diag_grad)(df_dx, x)
```

Result:
```
UserWarning: torch.vmap is an experimental prototype that is subject to change and/or deletion. Please use at your own risk. There may be unexpected performance cliffs due to certain operators not being implemented. To see detailed performance warnings please use `torch._C._debug_only_display_vmap_fallback_warnings(True) before the call to `vmap`.
  d2f_dx2 = torch.vmap(diag_grad)(df_dx, x)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-29-b92d38b59d51> in <module>
      7 def diag_grad(grad_i, x_i):
      8     return torch.autograd.grad(grad_i, x_i)
----> 9 d2f_dx2 = torch.vmap(diag_grad)(df_dx, x)
     10 print(jacobian)

~\miniconda3\lib\site-packages\torch\_vmap_internals.py in wrapped(*args)
    256         try:
    257             batched_inputs, batch_size = _create_batched_inputs(in_dims, args, vmap_level, func)
--> 258             batched_outputs = func(*batched_inputs)
    259             _validate_outputs(batched_outputs, func)
    260             return _unwrap_batched(batched_outputs, out_dims, vmap_level, batch_size, func)

<ipython-input-29-b92d38b59d51> in diag_grad(grad_i, x_i)
      6 
      7 def diag_grad(grad_i, x_i):
----> 8     return torch.autograd.grad(grad_i, x_i)
      9 d2f_dx2 = torch.vmap(diag_grad)(df_dx, x)
     10 print(jacobian)

~\miniconda3\lib\site-packages\torch\autograd\__init__.py in grad(outputs, inputs, grad_outputs, retain_graph, create_graph, only_inputs, allow_unused)
    221         retain_graph = create_graph
    222 
--> 223     return Variable._execution_engine.run_backward(
    224         outputs, grad_outputs_, retain_graph, create_graph,
    225         inputs, allow_unused, accumulate_grad=False)

RuntimeError: torch.autograd.grad(outputs, inputs, grad_outputs) called inside torch.vmap. We do not support the case where any outputs are vmapped tensors (output 0 is being vmapped over). Please call autograd.grad() outside torch.vmap or file a bug report with your use case.
```

## Expected behavior
It should output the correct second order gradients. Other methods to accomplish this are:
A loop implementation (slow due to loop):
```python
N = 5
f = lambda x: (x**2).sum()
x = torch.randn(N, requires_grad=True)
x_splitted = x.split(1, -1)
x = torch.cat(x_splitted)
y = f(x)
df_dx = torch.autograd.grad(y.sum(), x, create_graph=True)[0]

d2f_d2x = []
for g_i, x_i in zip(df_dx.split(1, -1), x_splitted):
    d2f_d2x.append(torch.autograd.grad(g_i, x_i, retain_graph=True))
```
A Jacobian implementation (slow due to O(N²)):
```python
N = 5
f = lambda x: (x**2).sum()
x = torch.randn(N, requires_grad=True)
y = f(x)
def df_dx(x):
    return torch.autograd.grad(f(x), x, create_graph=True)[0]

torch.autograd.functional.jacobian(df_dx, x)
```

## Environment
 - PyTorch Version (e.g., 1.0): 1.9.0.dev20210312
 - OS (e.g., Linux): Windows 10
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Build command you used (if compiling from source):
 - Python version: 3.7,5
 - CUDA/cuDNN version: 10.2
 - GPU models and configuration: RTX 2070
 - Any other relevant information:

**Comment by zou3519** [comment]
@n-gao out of curiosity, what is your use case for computing the Laplacian?

>  A torch.vmap implementation would only require O(N) time and space.

I have to think about this a little. Based on how autograd works, I'm not convinced that this would give an O(N) time and space way to compute the diagonal of the hessian. If we replace the `vmap` call in your example with a for-loop, then the example doesn't actually compute the correct quantity:
```
N = 5
f = lambda x: (x**2).sum()
x = torch.randn(N, requires_grad=True)
y = f(x)
df_dx, = torch.autograd.grad(y.sum(), x, create_graph=True)

def diag_grad(grad_i, x_i):
    return torch.autograd.grad(grad_i, x_i)
d2f_dx2 = [diag_grad(df_dx[i], x[i]) for i in range(N)]
RuntimeError: One of the differentiated Tensors appears to not have been used in the graph. Set allow_un
sed=True if this is the desired behavior.
```

There is an option to use vmap in our `torch.autograd.functional.hessian` API in PyTorch 1.8: https://pytorch.org/docs/stable/autograd.html (see the `vectorize` flag). So if you wanted to compute the laplacian by computing the full hessian and taking the diagonal, that would be one way to go.

**Comment by n-gao** [comment]
Hi @zou3519 ,
Thank you for your response! I am happy to elaborate a bit.

In you code snippet you skipped 2 lines:
```python
x_splitted = x.split(1, -1)
x = torch.cat(x_splitted)
```
These two lines ensure that I can take the derivative to the individual inputs.
If you run the snippet as in the bug report your output will be:
```python
>>> import torch
>>> N = 5
>>> f = lambda x: (x**2).sum()
>>> x = torch.randn(N, requires_grad=True)
>>> x_splitted = x.split(1, -1)
>>> x = torch.cat(x_splitted)
>>> y = f(x)
>>> df_dx = torch.autograd.grad(y.sum(), x, create_graph=True)[0]
>>>
>>> d2f_d2x = []
>>> for g_i, x_i in zip(df_dx.split(1, -1), x_splitted):
...     d2f_d2x.append(torch.autograd.grad(g_i, x_i, retain_graph=True))
...
>>> d2f_d2x
[(tensor([2.]),), (tensor([2.]),), (tensor([2.]),), (tensor([2.]),), (tensor([2.]),)]
```

Computing the complete Hessian is very memory intensive. Also, regarding `torch.autograd.functional.hessian` it does not seem to support batches (for a NxM tensor as input, I get a NxMxNxM output tensor). nor can't it be vmapped-
Here an example if I'd set `N=1000` and add a batch size of `100`.
||`grad + functional.jacobian`|`loop`|
|---|---|---|
|Memory|1349MiB| 969MiB|

**Comment by zou3519** [comment]
Thanks for the descriptive reply! It clears up a lot.

> In you code snippet you skipped 2 lines:

Oops, thanks for pointing that out. This makes sense now.

>  Also, regarding torch.autograd.functional.hessian it does not seem to support batches (for a NxM tensor as input, I get a NxMxNxM output tensor). nor can't it be vmapped

Makes sense. So you would want to, for each vector in the batch, compute a hessian.

**Comment by n-gao** [comment]
> Makes sense. So you would want to, for each vector in the batch, compute a hessian.
Exactly, more specific, only the diagonal elements but I think one has to compute the complete Hessian to get them. Though, a bit more memory-friendly way would be appreciated (which is possible as outlined above).

**Comment by zou3519** [comment]
@n-gao how large is N generally in your use case? Do you generally want to do this on CPU or CUDA?

**Comment by n-gao** [comment]
@zou3519 generally as big as possible, it is intended for a milti-gpu setting and better memory efficiency reduces the number of GPUs.

**Comment by n-gao** [comment]
To emphasize the need for an efficient way of computing this:
In my use case, the largest batch size I can use with `torch.autograd.functional.jacobian` (for N=90) is 128. With my loop implementation, I can get to a batch size of 4096. I tested both on a GTX 1080 with 12GB VRAM.

In this case f(x) is not a simple function but a large neural network.

I would really appreciate if it would be possible to enable second-order grads in `torch.vmap` like `jax` does.

**Comment by zou3519** [comment]
@n-gao, how would you do this in JAX in an O(N) way? I one way to express this computation (that is not O(n)), you'd compute the hessian and take the diagonal of it (and then vmap over this computation to support batches).

**Comment by n-gao** [comment]
@zou3519 The loop implementation which I provided only requires O(N) memory. This loop can be parallelized in jax via vmap.

**Comment by zou3519** [comment]
@n-gao I'm not sure it is possible to parallelize the loop implementation (for the hessian diagonal computation) via vmap and only require O(N) memory. Intuitively vmap trade-offs memory usage for performance; if we're vectorizing N computations together that each involve a tensor of size (N,) then the result is a computation that involves a tensor of size (N, N). 

I was curious so I wrote up a sample implementation using vmap in JAX:
```
import jax
import jax.numpy as jnp

N = 5
f = lambda x: jnp.sum(x**2)
x = jnp.arange(N, dtype=jnp.float32)

jacobian = jax.jacrev(f)(x)

def hessian_diag_i(x, i):
    def grad_i(x):
        return jax.grad(f)(x)[i]

    return jax.grad(grad_i)(x)[i]

idxs = jnp.arange(N)

result = jax.vmap(hessian_diag_i, (None, 0))(x, idxs)
```
We can actually inspect the sizes of the tensors being used:

```
In [28]: [eqn.outvars[0].aval for eqn in jaxpr.jaxpr.eqns]
Out[28]:
[ShapedArray(float32[5]),
 ShapedArray(float32[5]),
 AbstractUnit(),
 ShapedArray(float32[5]),
 ShapedArray(float32[5]),
 ShapedArray(float32[5]),
 ShapedArray(float32[5]),
 AbstractUnit(),
 ShapedArray(float32[5]),
 ShapedArray(float32[5]),
 ShapedArray(bool[5]),
 ShapedArray(int32[5]),
 ShapedArray(int32[5]),
 ShapedArray(int32[5,1]),
 ShapedArray(int32[5,0]),
 ShapedArray(int32[5,1]),
 AbstractUnit(),
 ShapedArray(float32[5]),
 ShapedArray(float32[5,5]),
 ShapedArray(float32[5]),
 ShapedArray(int32[5,1]),
 ShapedArray(int32[5,2]),
 ShapedArray(float32[5,5]),
 ShapedArray(float32[5,5]),
 ShapedArray(float32[5,5]),
 ShapedArray(float32[5,5]),
 ShapedArray(float32[5,5]),
 ShapedArray(float32[5,5]),
 ShapedArray(bool[5]),
 ShapedArray(int32[5]),
 ShapedArray(int32[5]),
 ShapedArray(int32[5,1]),
 ShapedArray(int32[5,0]),
 ShapedArray(int32[5,1]),
 ShapedArray(int32[5,1]),
 ShapedArray(int32[5,2]),
 ShapedArray(float32[5])]
```
from this it looks like there are tensors of size (N, N), so the memory consumption is O(N^2) as hypothesized.

**Comment by zou3519** [comment]
But memory usage aside, it sounds like there is still a use case for using vmap to do compute a laplacian per-sample in the batch. E.g. if we have a batch of size (M, N), then the use canse is to use vmap to compute M laplacians.

**Comment by n-gao** [comment]
@zou3519 You're right, one always needs O(N²) if the output should be computed in parallel.

**Comment by n-gao** [comment]
Hi Zou,

This is the JAX function which I would like to implement in PyTorch https://github.com/deepmind/ferminet/blob/aade61b3d30883b3238d6b50c85404d0e8176155/ferminet/hamiltonian.py#L23 .
However, they don't use the complete Hessian to compute the Laplacian but instead iterate and only compute the relevant diagonal entries. They also rely on forward-mode gradients which at the moment are not available in PyTorch. By a combination of these features, it turns out that the JAX implementation is a lot faster than what I could implement in PyTorch (without vectorization) (~10 times faster).

An implementation with a vectorized Hessian is probably faster if the dimension if x is small. Though, for large x it is more efficient to iterate over the diagonal.
